src/core/REPL.pm

```diff
@@ -277,27 +282,118 @@ do {
             self.?teardown-line-editor;
         }
 
-        method eval(Mu \SELF, Mu \super, Mu \code, Mu \args, Mu \adverbs) {
-            try {
-                my &needs_more_input = adverbs<needs_more_input>;
-                CATCH {
-                    when X::Syntax::Missing {
-                        if $!multi-line-enabled && .pos == code.chars {
-                            return needs_more_input();
-                        }
+        method partial-eval(Mu \code, Mu \adverbs) {
+            my &needs_more_input = adverbs<needs_more_input>;
+
+            CATCH {
+                when X::Syntax::Missing {
+                    if $!multi-line-enabled && .pos == code.chars {
+                        return needs_more_input();
+                    } else {
                         .throw;
                     }
+                }
 
-                    when X::Comp::FailGoal {
-                        if $!multi-line-enabled && .pos == code.chars {
-                            return needs_more_input();
-                        }
+                when X::Comp::FailGoal {
+                    if $!multi-line-enabled && .pos == code.chars {
+                        return needs_more_input();
+                    } else {
                         .throw;
                     }
                 }
 
-                super.(SELF, code, |@(args), |%(adverbs))
             }
+
+            my $result := self.compiler.eval(code, |%(adverbs));
+
+            return $result;
+        }
+
+        method repl-eval($code, *%adverbs) {
+            my $needs_more_input = False;
+            %adverbs<needs_more_input> := sub () {
+                $needs_more_input = True;
+            };
+            my $result := self.partial-eval($code, %adverbs);
+            if $needs_more_input {
```

Comment:
This isn't a blocker for your work; just something to keep in mind for the future, but since `repl-eval`'s invocation is completely contained within the Rakudo project, we have more flexibility in how we indicate
"more input needed" to the caller rather than being forced to use a sentinel value.